common/kvstore/tablestore/table_store_builder.go

```diff
@@ -0,0 +1,402 @@
+		return mapstore.NewStore(), nil
+	default:
+		return nil, fmt.Errorf("unknown store type: %d", storeType)
+	}
+}
+
+// validateSchema loads/initiates the schema version in the metadata table.
+func validateSchema(metadataTable kvstore.Table) error {
+
+	schemaKey := []byte(metadataSchemaVersionKey)
+	onDiskSchemaBytes, err := metadataTable.Get(schemaKey)
+
+	if err != nil {
+		if !errors.Is(err, kvstore.ErrNotFound) {
+			return err
+		}
+
+		// This store is new, no on disk schema version exists.
+		onDiskSchemaBytes = make([]byte, 8)
+		binary.BigEndian.PutUint64(onDiskSchemaBytes, currentSchemaVersion)
+
+		err = metadataTable.Put(schemaKey, onDiskSchemaBytes)
+		if err != nil {
+			return fmt.Errorf("error setting schema version in metadata table: %w", err)
+		}
+
+		return nil
+	}
+
+	// Verify schema version.
+	onDiskSchema := binary.BigEndian.Uint64(onDiskSchemaBytes)
+	if onDiskSchema != currentSchemaVersion {
+		// In the future if we change schema versions, we may need to write migration code here.
+		return fmt.Errorf(
+			"incompatible schema version: code is at version %d, data on disk is at version %d",
+			currentSchemaVersion, onDiskSchema)
+	}
+
+	return nil
+}
+
+// This method adds and removes tables as needed to match the given list of tables. The table ID map is updated
+// to reflect the new state of the tables.
+func addAndRemoveTables(
+	base kvstore.Store,
+	metadataTable kvstore.Table,
+	namespaceTable kvstore.Table,
+	tableIDMap map[uint32]string,
+	currentTables []string) error {
+
+	// Determine which tables to keep and which to drop.
+	originalTablesSet := make(map[string]bool)
+	for _, table := range tableIDMap {
+		originalTablesSet[table] = true
+	}
+	newTablesSet := make(map[string]bool)
+	for _, table := range currentTables {
+		newTablesSet[table] = true
+	}
+	tablesToDrop := make([]string, 0, len(originalTablesSet))
+	for table := range originalTablesSet {
+		if !newTablesSet[table] {
+			tablesToDrop = append(tablesToDrop, table)
+		}
+	}
+	tablesToAdd := make([]string, 0, len(newTablesSet))
+	for table := range newTablesSet {
+		if !originalTablesSet[table] {
+			tablesToAdd = append(tablesToAdd, table)
+		}
+	}
+
+	// Drop tables that are not in the list.
+	reverseTableIDMap := make(map[string]uint32)
+	for tableName, tableID := range tableIDMap {
+		reverseTableIDMap[tableID] = tableName
+	}
+	for _, tableName := range tablesToDrop {
+		err := dropTable(base, metadataTable, namespaceTable, tableName, reverseTableIDMap[tableName])
+		if err != nil {
+			return fmt.Errorf("error dropping table %s: %w", tableName, err)
+		}
+		delete(tableIDMap, reverseTableIDMap[tableName])
+	}
+
+	if len(tablesToAdd) == 0 {
+		// No tables to add, bail out early.
+		return nil
+	}
+
+	// Determine the table IDs for the new tables to be added.
+	// We want to fill gaps prior to assigning new IDs.
+	newTableIDs := make([]uint32, 0, len(tablesToAdd))
+	nextID := uint32(0)
+	for i := 0; i < len(tablesToAdd); i++ {
```

Comment:
Slightly simpler/more readable way to express this code:
```
for len(newTableIDs) < len(tablesToAdd) {
	if _, alreadyUsed := tableIDMap[nextID]; !alreadyUsed {
		newTableIDs = append(newTableIDs, nextID)
	}
	nextID++
}
```

Reply:
Yes, that's a lot cleaner. Done.